File: Semestr3/Paradygmaty-Programowania/Laboratoria/lab7/lab7.ml

type 'a llist = LNil | LCons of 'a * 'a llist Lazy.t;;

let rec toLazyList xs =
match xs with
[] -> LNil
| h::t -> LCons(h, lazy (toLazyList t));;

let rec gen n = 
    LCons(n, lazy (gen (n+1)));;

let rec ltake (n, lxs) =
match (n, lxs) with
(0, _) -> []
| (_, LNil) -> []
| (n, LCons(x, lazy xf)) -> x::ltake(n-1, xf)
;;

let lWybierz xs m n = 
    if n <= 0 then LNil else
    let rec skipN xs n = 
        match (n, xs) with
        | (_, LNil) -> LNil
        | (m, _) when m <= 0 -> xs
        | (_, LCons(head, lazy tail)) -> skipN (tail) (n-1)
    in 
        let rec process xs = 
            match xs with
            | LCons(head, lazy tail) -> LCons (head, lazy (process (skipN (tail) (n-1))))
            | LNil -> LNil
        in process (skipN xs (m-1));;

(*test zwykła lista*)
let l1 = toLazyList [5;6;3;2;1];;
let test1 = ltake (3,(lWybierz l1 1 2));;

(*test nieskonczona lista*)
let l2 = gen 4
let test2 = ltake (5,(lWybierz l2 9 9));;

(*test lista stringów*)
let l3 = toLazyList ["a"; "b"; "c"; "d"];;
let test3 = ltake (5,(lWybierz l3 2 2));;

(*test pusta lista*)
let l4 = toLazyList [];;
let test4 = ltake (5,(lWybierz l4 2 2));;

(* test ujemne m*)
let l5 = toLazyList [1;2];;
let test5 = ltake (5,(lWybierz l5 (-3) 1));;

(*test n wieksze niz ilosc elementow w liscie*)
let l6 = toLazyList [1;2];;
let test6 = ltake (5,(lWybierz l6 0 10));;

(*test m wieksze niz ilosc elementów*)
let l7 = toLazyList [1;2];;
let test7 = ltake (5,(lWybierz l7 3 10));;

File: Semestr3/Paradygmaty-Programowania/Laboratoria/lab7/lab7.scala
object lab7 extends App {
  
  def ldzialanie[A] (xs1: LazyList[A], xs2: LazyList[A], func : (A,A) => A) : LazyList[A] = {
    (xs1, xs2) match
      case (h1 #:: t1, h2 #:: t2) => func (h1, h2) #:: ldzialanie (t1, t2, func)
      case (LazyList(), LazyList()) => LazyList()
      case (LazyList(), h2 #:: t2) => h2 #:: ldzialanie (LazyList(), t2, func)
      case (h1 #:: t1, LazyList()) => h1 #:: ldzialanie (t1, LazyList(), func)
  }

  println("test: ldzialanie(LazyList(1,2,3,4), LazyList(1,2,3), _ + _) ")
  val test1 = ldzialanie(LazyList(1,2,3,4), LazyList(1,2,3), _ + _);
  println(test1.toList.toString());
  println("test2: ldzialanie(LazyList(a,b,c), LazyList(A,B,C), _ + _) ")
  val test2 = ldzialanie(LazyList("a", "b", "c"), LazyList("A", "B", "C"), _ + _ )
  println(test2.toList.toString())
  println("test3: ldzialanie(LazyList(1,2,3), LazyList(), (a: Int, b: Int) => a - b) ")
  val test3 = ldzialanie(LazyList(1,2,3), LazyList(), (a: Int, b: Int) => a - b)
  println(test3.toList.toString)
  println("test4: ldzialanie(LazyList(), LazyList(), (a: Int, b: Int) => a * b) ")
  val test4 = ldzialanie(LazyList(), LazyList(), (a:Int, b:Int) => a * b)
  println(test4.toList.toString)
  println("test5:val test5 = ldzialanie(LazyList.from(1), LazyList.from(1), (a:Int, b:Int) => a+b+1)")
  val test5 = ldzialanie(LazyList.from(1), LazyList.from(1), (a:Int, b:Int) => a/b)
  println(test5.take(10).toList.toString())

}
